knowledge/knowledge_read.go

```diff
@@ -23,195 +23,205 @@ import (
-// by its externalID + type combination.
-func (c *Client) GetResourceByIdentifier(
-	ctx context.Context,
-	identifier *Identifier,
-	opts ...grpc.CallOption,
-) (*knowledgepb.Node, error) {
-	ctx = insertMetadata(ctx, c.xMetadata)
-	return c.getNodeByIdentifier(ctx, Resource, identifier, opts...)
-}
-
-// ListDigitalTwinsByProperty is a helper function that lists all DigitalTwin nodes
+// ListIdentitiesByProperty is a helper function that lists all Identity nodes
 // that have the specified property.
-func (c *Client) ListDigitalTwinsByProperty(
+func (c *Client) ListIdentitiesByProperty(
 	ctx context.Context,
-	property *knowledgepb.Property,
-	opts ...grpc.CallOption) ([]*knowledgepb.Node, error) {
-	return c.ListNodesByProperty(ctx, DigitalTwin, property, opts...)
+	property *knowledgeobjects.Property,
+	returns []*knowledgepb.Return,
+	opts ...grpc.CallOption) ([]*knowledgeobjects.Node, error) {
+	return c.ListNodesByProperty(ctx, property, returns, true, opts...)
 }
 
-// ListResourcesByProperty is a helper function that lists all Resource nodes.
-// that have the specified property.
-func (c *Client) ListResourcesByProperty(
+// ListIdentities is a helper function that lists all Identity nodes.
+func (c *Client) ListIdentities(
 	ctx context.Context,
-	property *knowledgepb.Property,
-	opts ...grpc.CallOption) ([]*knowledgepb.Node, error) {
-	return c.ListNodesByProperty(ctx, Resource, property, opts...)
+	returns []*knowledgepb.Return,
+	opts ...grpc.CallOption) ([]*knowledgeobjects.Node, error) {
+	return c.ListNodes(ctx, returns, true, opts...)
 }
 
-// ListDigitalTwins is a helper function that lists all DigitalTwin nodes.
-func (c *Client) ListDigitalTwins(
-	ctx context.Context,
-	opts ...grpc.CallOption) ([]*knowledgepb.Node, error) {
-	return c.ListNodes(ctx, DigitalTwin, opts...)
-}
-
-// ListResources is a helper function that lists all Resource nodes.
-// that have the specified property.
-func (c *Client) ListResources(
-	ctx context.Context,
-	opts ...grpc.CallOption) ([]*knowledgepb.Node, error) {
-	return c.ListNodes(ctx, Resource, opts...)
-}
-
-// ListNodes is a helper function that lists all nodes by type, regardless of whether they are DigitalTwins
-// or Resources. The nodeType argument should be in PascalCase.
+// ListNodes is a helper function that lists all nodes by type, regardless of whether they are Identities
+// or not.
 func (c *Client) ListNodes(
 	ctx context.Context,
-	nodeType string,
+	returns []*knowledgepb.Return,
+	isIdentity bool,
 	opts ...grpc.CallOption,
-) ([]*knowledgepb.Node, error) {
+) ([]*knowledgeobjects.Node, error) {
 	ctx = insertMetadata(ctx, c.xMetadata)
-	path := fmt.Sprintf("(:%s)", nodeType)
-	resp, err := c.client.IdentityKnowledge(ctx, &knowledgepb.IdentityKnowledgeRequest{
-		Path:      path,
-		Operation: knowledgepb.Operation_OPERATION_READ,
+	query := "MATCH (n:Resource)"
+	if isIdentity {
+		query = fmt.Sprintf(
+			"%s WHERE n.is_identity=%t",
+			query,
+			isIdentity,
```

Comment:
all nodes that are identities will have the `DigitalTwin` label. So instead of this if, you can just change the query to `query := "MATCH (n:DigitalTwin)`

Reply:
updated